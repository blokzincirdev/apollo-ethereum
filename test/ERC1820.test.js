const gql = require('graphql-tag')

const FooInterface = artifacts.require('FooInterface')
const Policy = artifacts.require('Policy')

const config = {
  contracts: {
    FooInterface: {
      interfaceName: 'Foo',
    },
  },
}

const options = {
  erc1820: {
    lookupAddress: Policy.address,
    lookupMethod: 'policyFor',
  },
}

const QUERY = gql`
  query {
    FooInterface {
      _address
      implementsInterface
    }
  }
`

const MUTATION = gql`
  mutation {
    FooInterface {
      set(x: 12)
    }
  }
`

contract('ERC1820', () => {
  let client

  before(async () => {
    client = await createClient(config, { FooInterface }, options)
  })

  it('generates graphql schema', async () => {
    expect(client.source).to.matchSnapshot()
  })

  it('looks up address on query', async () => {
    const res = await client.execute(QUERY)
    expect(res.FooInterface.implementsInterface).to.equal('Foo')
    expect(res.FooInterface._address).to.equal(FooInterface.address)
  })

  it('looks up address on mutation', async () => {
    const res = await client.execute(MUTATION)
    expect(res.FooInterface.set).to.equal(true)
  })
})
