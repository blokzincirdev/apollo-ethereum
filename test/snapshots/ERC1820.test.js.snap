// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contract: ERC1820 generates graphql schema 1`] = `
"directive @erc1820(interfaceName: String) on FIELD_DEFINITION

scalar Address

scalar BigNumber

type FooInterface {
  _address: Address
  implementsInterface: String
}

type FooInterfaceMutative {
  set(x: BigNumber): Boolean
}

type Mutation {
  FooInterface(address: Address): FooInterfaceMutative @erc1820(interfaceName: \\"Foo\\")
}

type Query {
  FooInterface(address: Address): FooInterface @erc1820(interfaceName: \\"Foo\\")
}
"
`;
